tools/clustering/src/clustering.py

```diff
@@ -0,0 +1,269 @@
+
+    assert 0 < coverage <= 1, "Coverage should be within the (0,1] range."
+
+    n = trace.get_invocation_count()
+
+    print("Finding eps parameter based on coverage of %f..." % coverage)
+    distance_matrix = trace.get_distance_matrix(disjoint_sets)
+    distances = []
+    for i in range(n):
+        distances.append(min([distance_matrix[i, j] for j in range(n) if j != i]))
+
+    pointLimit = math.ceil((n - 1) * coverage)
+    distances = np.sort(distances)
+    uniquedist = np.unique(distances)
+    distanceLimit = distances[pointLimit]
+    nextDist = uniquedist[1 + np.where(uniquedist == distanceLimit)[0]]
+
+    print("eps parameter based on coverage of %f set to %f" % (coverage, (distanceLimit + nextDist)/2))
+    return (distanceLimit + nextDist)/2
+
+def dbscan(trace, epsilon):
+    distance_matrix = trace.get_distance_matrix(disjoint_sets)
+
+    print("Clustering using parameters: eps = %f;" % epsilon)
+    db = DBSCAN(metric="precomputed", eps=epsilon).fit(distance_matrix)
+    core_samples_mask = np.zeros_like(db.labels_, dtype=bool)
+    core_samples_mask[db.core_sample_indices_] = True
+    labels = db.labels_
+
+    # Number of clusters in labels, ignoring noise if present.
+    n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)
+    n_noise_ = list(labels).count(-1)
+
+    print('Estimated number of clusters: %d' % n_clusters_)
+    print('Estimated number of noise points: %d (%f%%)' % (n_noise_, n_noise_*100/trace.get_invocation_set_count()))
+    if n_clusters_ > 1:
+        silhouette_score = metrics.silhouette_score(distance_matrix, labels,
+                                                    metric="precomputed")
+        print("Silhouette Coefficient: %0.3f" % silhouette_score)
+
+    clusters = [[] for i in range(n_clusters_)]
+    noise_invocations = []
+    for i in range(len(labels)):
+        label = labels[i]
+
+        if label == -1:
+            noise_invocations.append(i)
+        else:
+            clusters[label].append(i)
+
+    invocation_sets = [invocation_set.invocations for invocation_set in trace.invocation_sets]
+
+    return ClusteringInformation(epsilon, invocation_sets, clusters, noise_invocations)
+
+def dbscan_ipc_instr(invocations, epsilon):
+    data = []
+    for i in range(len(invocations)):
+        invocation = invocations[i]
+        done = False
+        for dp in data:
+            if not done and abs(1 - (invocation.metrics.instructions / dp[1])) < 0.01:
+                count = len(dp[0])
+                dp[0].append(invocation)
+                dp[1] = ((dp[1] * count) + invocation.metrics.instructions) / (count + 1)
+                dp[2] = ((dp[2] * count) + invocation.metrics.ipc) / (count + 1)
+                dp[3].append(i)
+                done = True
+        if done == False:
+            data.append([[invocation], invocation.metrics.instructions, invocation.metrics.ipc, [i]])
+
+    X = np.array([[dp[2] * 20, math.log(dp[1])] for dp in data])
```

Comment:
@ArnauBigas Here, I assume that you try to normalize the range of both metrics: ipc and # of instructions. What is the reason to use 20 and log? Is this something you _tunned_ experimentally, if so, we need to expose such "normalization" options to the user. Maybe other experiments, require other _tunning_ ...

That said, I'm of the opinion of using ageneric approach. Can we use data normalization and/or standardization on so that both metrics have the same _weight_? [1] . Maybe we already discussed this in the past and the results were not satisfactory(?). In any case, if that is considered, we need to remove outliers on the data before doing the transformation (and assign such data points directly as noise), otherwise the clustering may be affected. 

[1] https://python-data-science.readthedocs.io/en/latest/normalisation.html

Reply:
You are mostly right. It was tuned experimentally, based on the intuition that IPC would have a range of (0, 4?), while the retired instruction count would be in the millions or even billions. Thus, the reason for the log() and the smaller adjustment on the IPC.

After having had some more university courses on this subject since I wrote this, maybe we could try using a standard scaler such as [minmax scaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html)? We would have to experimentally confirm it works, though.

Reply:
Yupp, scaling the range to [0,1] (after removing outliers) will be a more _generic_ solution since it will give both metrics (ipc and # instructions) the same _weight_. We can try the minmax scaler, but it may generate problems by changing the sparsity of the data. For example, in the case where we have only two big values (e.g. `[400000010 ,400000000]`), the transformation will be `[0,1]` and that will _misslead_ the clustering algorithm (the points are actually very close). I'm more inclined to use the  [MaxAbsScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MaxAbsScaler.html#sklearn.preprocessing.MaxAbsScaler), since it does not destroy any sparsity. Obviously, we need to try it experimentally and see if this is the case...

Reply:
That's a very good point, that scaler might work better.

However, regardless of whatever scaler we use, I'm worried about the "epsilon" parameter of the clustering algorithm (DBSCAN). This parameter "defines the minimum distance between two points of different clusters" (intuitively describing what I know the parameter effectively does). The parameter right now is user-defined, and hard-coded in my scripts. I'm not sure if we should try to find the optimum value for this parameter instead? This would induce a high overhead, though, and the hardcoded value works quite fine so far with the "experimentally tuned scaler"...

Reply:
I see, you already tuned the epsilon for the experiments done so far and we do not want to redo all that. So, we keep this clustering algorithm as is. We can implement modifications/improvements in a different clustering strategy (i.e. in a different function). 

Regarding about finding the optimum epsilon automatically, I'm totally in. My goal is to have a good enough _push button_ solution without requiring any use input/expertise (just the command ;) ) So, I'm thinking in something along the following lines: Perform a _greedy_ search of different eps at the expense of execution time. Start with a _strict_ eps (many clusters or many noise points), and iteratively modify eps until: (a) no more than X clusters are defined and (b) no more than %X of points are noise. Does that makes sense?